posts/2023-10-30_floating_point/floating_point.qmd

```diff
@@ -0,0 +1,338 @@
+long_slug <- "2023-10-30_floating_point"
+# renv::use(lockfile = "renv.lock")
+```
+
+<!--------------- post begins here ----------------->
+```{r, echo=FALSE}
+# check if version 444 of admiral is installed, and throw error if not:
+if (packageVersion("admiral") < "0.12.2") {
+  stop("Please install a more recent version of admiral.
+       At least version 0.12.2 is required.")
+}
+```
+
+```{r, echo=FALSE}
+```
+
+[{{admiral}}](https://github.com/pharmaverse/admiral) recently ran into some trouble when dealing with floating point values, captured [by this thread on GitHub](https://github.com/pharmaverse/admiral/pull/2060).
+This post gives a brief overview on floating point values, recaps the discussion on GitHub, and explains how [{{admiral}}](https://github.com/pharmaverse/admiral) deals with floating point values.
+
+## Floating point values
+
+Floating point values are numeric objects representing numbers between integers, e.g. 0.5., 2.3, 3.1415, etc.
+However, floating point numbers are not stored like integers, and most floating point numbers are approximations to the number they represent.
+To see what value a floating point number is actually stored as, we can use the `format()` function where we can increase the number of digits shown:
+
+```{r, echo=FALSE, message = FALSE}
+library(dplyr)
+format(1.4, digits = 22)
+```
+
+These very small numerical differences impact the result of mathematical operations:
+
+```{r}
+0.1 + 0.2 == 0.3
+```
+
+If we look at the actually stored values, this makes sense:
+
+```{r}
+0.1 %>% format(digits = 22)
+0.2 %>% format(digits = 22)
+
+(0.1 + 0.2) %>% format(digits = 22)
+
+0.3 %>% format(digits = 22)
+```
+
+The bottom line is: Avoid using exact comparators such as `==` and `>=` when comparing floating point values.
+
+::: callout-note
+
+## Exact floating point values
+
+There are some floating point values which can be exactly represented.
+All these values can be represented as $\frac{x}{2^y}$, where x and y are integers.
+For example, 0.5 is stored as $\frac{1}{2}$, 0.25 is stored as $\frac{1}{4}$, 0.125 is stored as $\frac{1}{8}$, etc.
+
+```{r}
+# simple examples
+0.5 %>% format(digits = 22)
+0.25 %>% format(digits = 22)
+0.125 %>% format(digits = 22)
+0.0625 %>% format(digits = 22)
+
+# some weird values for x and y
+(1121 / (2^9)) %>% format(digits = 22)
+```
+
+All floating point values are stored as $\frac{x}{2^y}$, where the outcome may be a very close approximation to the value they represent\*.
+
+\* Based on a recollection of the course associated with [this GitHub Repository](https://github.com/mmaechler/ProgRRR) by Martin Mächler.
+
+:::
+
+## Issues arising
+
+Gordon Miller stumbled upon this issue when he was creating a unit test for the `derive_var_atoxgr_dir()` function in [{{admiral}}](https://github.com/pharmaverse/admiral).
+This function made use of `atoxgr_criteria_ctcv5`, a `data-frame` object containing metadata on adverse events in cancer therapy.
```

Comment:
it was when I was creating DAIDS criteria that I came across issue. The case statement is quite long it might be easier to read in GRADE_1 variable on `atoxgr_criteria_daids`, and subset on TERMS, `Amylase, High`, `Creatinine, High` and `Lipase, High`. Then just mention we use a case statement to implement the grade. Doc for DAIDS is here: https://rsc.niaid.nih.gov/clinical-research-sites/daids-adverse-event-grading-tables